---
title: "MutationAnalysis"
author: "Theo-60985751, Hannah- , Jingxuan-"
date: "2023-11-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Including the libraries

```{r}
library(ggplot2)
library(pheatmap)
library("TCGAbiolinks")
library("survival")
library("survminer")
library("SummarizedExperiment")
library(DESeq2)
library("gridExtra")
library("AnnotationDbi")
library("org.Hs.eg.db")
```


# Opening the datafiles

```{r}
data_mutation = read.delim("data_mutations.txt", header = TRUE, sep = '\t')
data_clinical = read.table("data_clinical_patient.txt",header = TRUE, sep = '\t')
data_rna = read.csv("RNAseq_LIHC.csv", header = TRUE, row.names = "X")
```

# Finding the patients that we have data for clinical, genome, and RNAseq data

```{r}
unique_patient_clinical = unique(data_clinical$PATIENT_ID)
unique_patient_rna = unique(colnames(data_rna))
unique_patient_mutation = unique(data_mutation$Tumor_Sample_Barcode)

shortened_rna = substr(unique_patient_rna, start = 1, stop = 12)
shortened_rna = gsub("\\.", "-", shortened_rna)

shortened_mutation = substr(unique_patient_mutation, start = 1, stop = 12)

common_names1 <- intersect(unique_patient_clinical, shortened_mutation)
common_names2 <- intersect(common_names1, shortened_rna)

```

###  Adding the patient ID at the last column and get the common datas
```{r}
data_clinical_common = subset(data_clinical, PATIENT_ID %in% common_names2)

data_mutation$PATIENT_ID = substr(data_mutation$Tumor_Sample_Barcode, start = 1, stop = 12)
data_mutation_common = subset(data_mutation, PATIENT_ID %in% common_names2)

data_mutation_common$Tumor_Sample_Barcode = data_mutation_common$PATIENT_ID

data_rna_shortened = data_rna
colnames(data_rna_shortened) = substr(colnames(data_rna_shortened), start = 1, stop = 12)
colnames(data_rna_shortened) = gsub("\\.", "-", colnames(data_rna_shortened))
data_rna_common = data_rna_shortened[,common_names2]

```

# Making a matrix of column names = patient ID, row names = genes



### From tutorial 
### Generate oncoplot matrix

```{r}
data_oncoplot = data_mutation_common

hugo <- as.data.frame(table(data_oncoplot$Hugo_Symbol))
var.class <- as.data.frame(table(data_oncoplot$Variant_Classification))
ggplot(data=var.class, aes(x=Var1, y=Freq))+
  geom_col()+
  theme(axis.text.x = element_text(angle = 45,hjust=1))
```

```{r}
hugo <- as.data.frame(table(data_mutation_common$Hugo_Symbol))

hugo.ordered <- hugo[order(-hugo$Freq),]

ggplot(data=hugo.ordered[1:20,], aes(x=Var1, y=Freq))+
  geom_col()+
  theme(axis.text.x = element_text(angle = 45,hjust=1))+
  scale_x_discrete(limits = hugo.ordered[1:20,]$Var1)
```


```{r}
cnv_events = unique(data_oncoplot$Variant_Classification)
oncomat = reshape2::dcast(
  data = data_oncoplot,
  formula = Hugo_Symbol ~ Tumor_Sample_Barcode,
  fun.aggregate = function(x, cnv = cnv_events) {
    x = as.character(x) # >= 2 same/distinct variant classification = Multi_Hit
    xad = x[x %in% cnv]
    xvc = x[!x %in% cnv]
    
    if (length(xvc) > 0) {
      xvc = ifelse(test = length(xvc) > 1,
                   yes = 'Multi_Hit',
                   no = xvc)
    }
    
    x = ifelse(
      test = length(xad) > 0,
      yes = paste(xad, xvc, sep = ';'),
      no = xvc
    )
    x = gsub(pattern = ';$',
             replacement = '',
             x = x)
    x = gsub(pattern = '^;',
             replacement = '',
             x = x)
    return(x)
  },
  value.var = 'Variant_Classification',
  fill = '',
  drop = FALSE
)
```



### Modify the row label to reflect the gene name
```{r}

rownames(oncomat) = oncomat$Hugo_Symbol
oncomat <- oncomat[,-1]
```



### Reorder the rows according to the occurrence of heavily mutated genes

```{r}
oncomat.ordered <- oncomat[order(-hugo$Freq),]
```


### Transform the matrix into a binary matrix
```{r}
mat <- oncomat.ordered
mat[mat== "Silent"] = 0
mat[mat == "Intron"] = 0
mat[mat == "Missense_Mutation"] = 0
mat[mat == ""] = 0



mat <- apply(mat, 2 ,as.numeric)
mat <- as.matrix(mat)
mat[is.na(mat)]=1

rownames(mat)  <-  row.names(oncomat.ordered)
```
### Finding the top 20 mutated genes
```{r}
genes = rowSums(mat)
oncomat.ordered <- oncomat[order(-hugo$Freq),]
genes.ordered = sort(genes, decreasing = TRUE)


genes.ordered.top = genes.ordered
genes.ordered.top.names = names(genes.ordered.top)

data_oncoplot.top = subset(data_oncoplot, Hugo_Symbol %in% genes.ordered.top.names)



```

### Making a matrix of the top 20 occuring genes

```{r}
cnv_events = unique(data_oncoplot.top$Variant_Classification)
oncomat.top = reshape2::dcast(
  data = data_oncoplot.top,
  formula = Hugo_Symbol ~ Tumor_Sample_Barcode,
  fun.aggregate = function(x, cnv = cnv_events) {
    x = as.character(x) # >= 2 same/distinct variant classification = Multi_Hit
    xad = x[x %in% cnv]
    xvc = x[!x %in% cnv]
    
    if (length(xvc) > 0) {
      xvc = ifelse(test = length(xvc) > 1,
                   yes = 'Multi_Hit',
                   no = xvc)
    }
    
    x = ifelse(
      test = length(xad) > 0,
      yes = paste(xad, xvc, sep = ';'),
      no = xvc
    )
    x = gsub(pattern = ';$',
             replacement = '',
             x = x)
    x = gsub(pattern = '^;',
             replacement = '',
             x = x)
    return(x)
  },
  value.var = 'Variant_Classification',
  fill = '',
  drop = FALSE
)
hugo <- as.data.frame(table(data_oncoplot.top$Hugo_Symbol))

rownames(oncomat.top) = oncomat.top$Hugo_Symbol
oncomat.top <- oncomat.top[,-1]
oncomat.top.ordered <- oncomat.top[order(-hugo$Freq),]



```

### Transforming the top matrix into binary

```{r, echo = FALSE}
mat.top <- oncomat.top.ordered
mat.top[mat.top== "Silent"] = 0
mat.top[mat.top == "Intron"] = 0
mat.top[mat.top == "Missense_Mutation"] = 0

mat.top[mat.top == ""] = 0


mat.top <- apply(mat.top, 2 ,as.numeric)
mat.top <- as.matrix(mat.top)
mat.top[is.na(mat.top)]=1

rownames(mat.top)  <-  row.names(oncomat.top.ordered)
```


```{r}

reduce.mat <- mat.top[1:2,]
res <- pheatmap(reduce.mat,
         cluster_rows = F,
         show_colnames = FALSE)
```
### clustering
```{r}
cluster = as.data.frame(cutree(res$tree_col, k = 2))
cluster
```
### Finding the patientID in each group
```{r}
mutation_group1_patientID = subset(cluster,`cutree(res$tree_col, k = 2)` == 1)
mutation_group2_patientID = subset(cluster,`cutree(res$tree_col, k = 2)` == 2)

data_clinical_common_group1 = subset(data_clinical_common,PATIENT_ID %in% rownames(mutation_group1_patientID))

data_clinical_common_group2 = subset(data_clinical_common,PATIENT_ID %in% rownames(mutation_group2_patientID))

mutation_TP53_patientID = data_mutation_common$PATIENT_ID[data_mutation_common$Hugo_Symbol == "TP53"]

data_clinical_common_TP53 = subset(data_clinical_common, PATIENT_ID %in% mutation_TP53_patientID)

```


### survival analysis on clinical data 
```{r}
data_clinical_common$deceased = data_clinical_common$PFS_STATUS == "1:PROGRESSION"

# create an "overall survival" variable that is equal to days_to_death
# for dead patients, and to days_to_last_follow_up for patients who
# are still alive
data_clinical_common$overall_survival = ifelse(data_clinical_common$deceased,
                                   data_clinical_common$OS_MONTHS,
                                   data_clinical_common$DAYS_LAST_FOLLOWUP)
```
### SA by sex
```{r}
Surv(data_clinical_common$overall_survival, data_clinical_common$deceased) ~ data_clinical_common$SEX

table(data_clinical_common$SEX)
fit = survfit(Surv(overall_survival, deceased) ~ SEX, data = data_clinical_common)

pval = surv_pvalue(fit, data=data_clinical_common)$pval
print(pval)
```
```{r}
ggsurvplot(fit, data=data_clinical_common, pval=T, risk.table=T, risk.table.col="strata", risk.table.height=0.35)
```

### SA by tumor stage
```{r}
# remove any of the letters "a", "b" or "c", but only if they are at the end
# of the name, eg "stage iiia" would become simply "stage iii"
data_clinical_common$AJCC_PATHOLOGIC_TUMOR_STAGE = gsub("[ABC]$", "", data_clinical_common$AJCC_PATHOLOGIC_TUMOR_STAGE)


table(data_clinical_common$AJCC_PATHOLOGIC_TUMOR_STAGE)
```

```{r}
fit = survfit(Surv(overall_survival, deceased) ~ AJCC_PATHOLOGIC_TUMOR_STAGE, data=data_clinical_common)

# we can extract the survival p-value and print it
pval = surv_pvalue(fit, data=data_clinical_common)$pval
print(pval)
```

```{r}
ggsurvplot(fit, data=data_clinical_common, pval=T, risk.table=T, risk.table.height=0.45)
```
expression frame done

### Survival analysis on the clustered patients

SA on cluster 1 Tumor Stage 
```{r}
data_clinical_common_group1$deceased = data_clinical_common_group1$PFS_STATUS == "1:PROGRESSION"

# create an "overall survival" variable that is equal to days_to_death
# for dead patients, and to days_to_last_follow_up for patients who
# are still alive
data_clinical_common_group1$overall_survival = ifelse(data_clinical_common_group1$deceased,
                                   data_clinical_common_group1$OS_MONTHS,
                                   data_clinical_common_group1$DAYS_LAST_FOLLOWUP)

# remove any of the letters "a", "b" or "c", but only if they are at the end
# of the name, eg "stage iiia" would become simply "stage iii"
data_clinical_common_group1$AJCC_PATHOLOGIC_TUMOR_STAGE = gsub("[ABC]$", "", data_clinical_common_group1$AJCC_PATHOLOGIC_TUMOR_STAGE)


table(data_clinical_common_group1$AJCC_PATHOLOGIC_TUMOR_STAGE)

fit = survfit(Surv(overall_survival, deceased) ~ AJCC_PATHOLOGIC_TUMOR_STAGE , data=data_clinical_common_group1)

# we can extract the survival p-value and print it
pval = surv_pvalue(fit, data=data_clinical_common_group1)$pval
print(pval)

ggsurvplot(fit, data=data_clinical_common_group1, pval=T, risk.table=T, risk.table.col="strata", risk.table.height=0.35)
```
SA on cluster 1 Sex
```{r}
Surv(data_clinical_common_group1$overall_survival, data_clinical_common_group1$deceased) ~ data_clinical_common_group1$SEX

fit = survfit(Surv(overall_survival, deceased) ~ SEX, data = data_clinical_common_group1)

table(data_clinical_common_group1$SEX)

pval = surv_pvalue(fit, data=data_clinical_common_group1)$pval
print(pval)
ggsurvplot(fit, data=data_clinical_common_group1, pval=T, risk.table=T, risk.table.col="strata", risk.table.height=0.35)
```

SA on cluster 2 Tumor stage 
```{r}
data_clinical_common_group2$deceased = data_clinical_common_group2$PFS_STATUS == "1:PROGRESSION"

# create an "overall survival" variable that is equal to days_to_death
# for dead patients, and to days_to_last_follow_up for patients who
# are still alive
data_clinical_common_group2$overall_survival = ifelse(data_clinical_common_group2$deceased,
                                   data_clinical_common$OS_MONTHS,
                                   data_clinical_common$DAYS_LAST_FOLLOWUP)

# remove any of the letters "a", "b" or "c", but only if they are at the end
# of the name, eg "stage iiia" would become simply "stage iii"
data_clinical_common_group2$AJCC_PATHOLOGIC_TUMOR_STAGE = gsub("[ABC]$", "", data_clinical_common_group2$AJCC_PATHOLOGIC_TUMOR_STAGE)


table(data_clinical_common_group2$AJCC_PATHOLOGIC_TUMOR_STAGE)

fit = survfit(Surv(overall_survival, deceased) ~ AJCC_PATHOLOGIC_TUMOR_STAGE, data=data_clinical_common_group2)

# we can extract the survival p-value and print it
pval = surv_pvalue(fit, data=data_clinical_common_group2)$pval
print(pval)

ggsurvplot(fit, data=data_clinical_common_group2, pval=T, risk.table=T, risk.table.col="strata", risk.table.height=0.35)
```
SA on cluster 2 Sex
```{r}
Surv(data_clinical_common_group2$overall_survival, data_clinical_common_group2$deceased) ~ data_clinical_common_group2$SEX

fit = survfit(Surv(overall_survival, deceased) ~ SEX, data = data_clinical_common_group2)

table(data_clinical_common_group2$SEX)
pval = surv_pvalue(fit, data=data_clinical_common_group2)$pval
print(pval)

ggsurvplot(fit, data=data_clinical_common_group2, pval=T, risk.table=T, risk.table.col="strata", risk.table.height=0.35)
```

SA on the most mutated gene sex (TP53)
```{r}
data_clinical_common_TP53$deceased = data_clinical_common_TP53$PFS_STATUS == "1:PROGRESSION"

# create an "overall survival" variable that is equal to days_to_death
# for dead patients, and to days_to_last_follow_up for patients who
# are still alive
data_clinical_common_TP53$overall_survival = ifelse(data_clinical_common_TP53$deceased,
                                   data_clinical_common_TP53$OS_MONTHS,
                                   data_clinical_common_TP53$DAYS_LAST_FOLLOWUP)


Surv(data_clinical_common_TP53$overall_survival, data_clinical_common_TP53$deceased) ~ data_clinical_common_TP53$SEX

fit = survfit(Surv(overall_survival, deceased) ~ SEX, data = data_clinical_common_TP53)

table(data_clinical_common_TP53$SEX)
pval = surv_pvalue(fit, data=data_clinical_common_TP53)$pval
print(pval)

ggsurvplot(fit, data=data_clinical_common_TP53, pval=T, risk.table=T, risk.table.col="strata", risk.table.height=0.35)
```
SA on the most mutated gene tumor stage
```{r}

# remove any of the letters "a", "b" or "c", but only if they are at the end
# of the name, eg "stage iiia" would become simply "stage iii"
data_clinical_common_TP53JCC_PATHOLOGIC_TUMOR_STAGE = gsub("[ABC]$", "", data_clinical_common_TP53$AJCC_PATHOLOGIC_TUMOR_STAGE)


table(data_clinical_common_TP53$AJCC_PATHOLOGIC_TUMOR_STAGE)

fit = survfit(Surv(overall_survival, deceased) ~ AJCC_PATHOLOGIC_TUMOR_STAGE, data=data_clinical_common_TP53)

# we can extract the survival p-value and print it
pval = surv_pvalue(fit, data=data_clinical_common_TP53)$pval
print(pval)

ggsurvplot(fit, data=data_clinical_common_TP53, pval=T, risk.table=T, risk.table.col="strata", risk.table.height=0.35)
```


### DE analysis
```{r}
data_rna_common <- data_rna_common[rowSums(data_rna_common)>1,]
head(data_rna_common)

```

```{r}
sampleDists = dist(t(data_rna_common),upper = TRUE)
# sampleDists
```


```{r}
# annot_col = data.frame(colData$condition)
# row.names(annot_col) <- rownames(colData)

sampleDistMatrix = as.matrix( sampleDists )
rownames(sampleDistMatrix) = colnames(data_rna_common)
colnames(sampleDistMatrix) = colnames(data_rna_common)

pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE,
         annotation_col=cluster)
```
PCA plot

```{r}
pca_res <- prcomp(t(data_rna_common), scale. = TRUE)
score <- pca_res$x

score = as.data.frame(score)
score$color <- as.factor(cluster$`cutree(res$tree_col, k = 2)`)

ggplot(score, aes(x=PC1, y=PC2,  color=color)) + 
  geom_point(size = 4)
```

For visualization purposes

```{r}
p1 <- ggplot(as.data.frame(data_rna_common),aes(x="TCGA-2V-A95S"))+
  geom_histogram(bins=50)+
  labs(title="No transformation")

log2 <- log(data_rna_common+1, base = 2)
p2 <- ggplot(as.data.frame(log2),aes(x="TCGA-2V-A95S"))+
  geom_histogram(bins=50)+
  labs(title="Log(x+1)")

# grid.arrange(p1, p2, ncol = 2)
```

Actually running the DESeq pipeline

```{r}
# dds = DESeqDataSetFromMatrix(countData=data_rna_common,
#                               colData=cluster,
#                               design= ~cluster$`cutree(res$tree_col, k = 2)`)
```


```{r}
rownames(cluster) = substr(rownames(cluster), start = 1, stop = 12)
common_cluster <- row.names(cluster)

data_rna_common_cluster <- data_rna_common[,common_cluster]

metadata <- data.frame(
  patientID = row.names(cluster),
  condition = cluster$`cutree(res$tree_col, k = 2)`
)
```


### DE

```{r}
dds <- DESeqDataSetFromMatrix(countData = data_rna_common_cluster, colData = metadata, design = ~ condition)
dds = DESeq(dds)
dds

# res_de <- results(dds)
# res_de
# summary(res_de)
# res_de.05 <- results(dds, alpha = 0.05)
# table(res_de.05$padj < 0.05)
```
```{r}
res_de = results(dds, contrast=c(1, 2))
mcols(res_de, use.names = TRUE)
summary(res_de)
res_de.05 <- results(dds, alpha = 0.05)
table(res_de.05$padj < 0.05)

resLFC1 <- results(dds, lfcThreshold=1)
table(resLFC1$padj < 0.1)
```


P-values
```{r}
res_de <- res_de[order(res_de$pvalue),]
summary(res_de)

## number of adjusted p-values less than 0.1
sum(res_de$padj < 0.1, na.rm=TRUE)
```

multiple testing

```{r}
sum(res_de$pvalue < 0.05, na.rm=TRUE)
sum(!is.na(res_de$pvalue))
sum(res_de$padj < 0.05, na.rm=TRUE)

resSig <- subset(res_de, padj < 0.05)
head(resSig[ order( resSig$log2FoldChange ), ])

head(resSig[ order( resSig$log2FoldChange, decreasing=TRUE), ])
```

MA-plot
```{r}
plotMA(res_de, ylim=c(-2,2))
```

Plot counts
```{r}
plotCounts(dds, gene=which.min(res_de$padj), intgroup="condition")
```
Effect of transformations on the variance (CANT RUN THE CHUNK BELOW)

```{r}
# this gives log2(n + 1)
ntd <- normTransform(dds)
# Variance stabilizing transformation
vsd <- vst(dds)

# Regularized log transformation
# The blind=TRUE argument results in a transformation unbiased to sample condition information.
rld <- rlog(dds, blind=FALSE)
```


```{r}
sampleDists = dist(t(assay(rld)),upper = TRUE)

# annot_col = data.frame(cluster$`cutree(res$tree_col, k = 2)`)
# row.names(annot_col) <- rownames(clusters)

sampleDistMatrix = as.matrix( sampleDists )
rownames(sampleDistMatrix) = colnames(data_rna_common)
colnames(sampleDistMatrix) = colnames(data_rna_common)

pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE,
         annotation_col=cluster)
```

```{r}
pca_res <- prcomp(t(assay(vsd)), scale. = TRUE)
score <- pca_res$x

score = as.data.frame(score)
score$color <- as.factor(cluster$`cutree(res$tree_col, k = 2)`)


ggplot(score, aes(x=PC1, y=PC2,  color=color)) + 
  geom_point(size = 4)
```

```{r}
genes <- order(res$padj, decreasing = TRUE) [1:20]

v kj        c
```


H0: There's no significance between cluster 1 and 2
H1: There's a difference between cluster 1 and 2

First: make a matrix with column names = patient_ID, row names = gene names,
values = 1 or 0.

Get all the genes with 1, then pick out the top 20 occuring genes (or play around with the top numbers)

Make another matrix with rownames = the top 20 occuring genes, 

make survival analysis (see if clinical data is different when compare to patient clusters)

DE analysis (see if any correlation between the clusters from mutation data and expression data)
p adjusted value, extract the significant gene, order them
